application/Common/HookLibrary/dllmain.h

```diff
@@ -0,0 +1,150 @@
+/**
+    This DLL allows hooking into processes' windows and retreiving
+    information on messages sent to them.
+*/
+
+#ifndef DLLMAIN_H
+#define DLLMAIN_H
+#include "pch.h"
+#include <Windows.h>
+#include <thread>
+#include <synchapi.h>
+
+/**
+    A serializable structure used for sending message data to the MORR application.
+ */
+typedef struct {
+    UINT32 Type;
+    HWND Hwnd;
+    WPARAM wParam;
+    INT32 data[4];
+} WM_Message;
+
+/**
+    Helper definitions for exported and retrieved functions.
+ */
+#define DLL extern "C" __declspec(dllexport)
+typedef void(__stdcall* WH_MessageCallBack)(WM_Message);
+
+/**
+    Size of the message buffer.
+ */
+#define BUFFERSIZE 2048
+ /**
+     Identifier of the inter-process semaphore.
+  */
+#define SEMAPHORE_GUID "7c4db072-3baf-457f-8259-da0c369e3ec8"
+
+#define MESSAGETABLESIZE 1025
+#pragma data_seg("Shared")
+/**
+    Shared data segment accessible by all injected applications and MORR itself.
+ */
+
+/**
+    Hook handles.
+ */
+HHOOK GetMessageHook = NULL;
+HHOOK CallWndProcHook = NULL;
+
+/**
+    The iterator determining where in the buffer to store the next message.
+*/
+UINT globalBufferIterator = 0;
+
+/**
+    The callback function to be invoked inside the MORR address space to send WM_Message structs
+    to the C# application.
+ */
+WH_MessageCallBack globalCallback;
+
+/**
+    Ringbuffers for the stored messages and their timestamps.
+    The latter is a helper array which won't get productively.
+    When using synchronization techniques in the message-loop,
+    events sometimes get duplicated. These timestamps are used
+    to identify if the currently inspected event is a duplicate
+    of the last one by matching their exact timestamps.
+ */
+DWORD globalTimeStamps[BUFFERSIZE] = { 0 };
+WM_Message globalMessageBuffer[BUFFERSIZE] = { 0 };
+
+/**
+    An individual table entry has to be set to true if any listener wants this message information.
+    Used to prevent unnecessary processing of undesired messages.
+ */
+bool messageHasListener[MESSAGETABLESIZE] = { 0 };
```

Comment:
Same argumentation as with std::array. I will try this when the current (working?) solution is merged to avoid having to test it all over again before merging.